android/app/src/main/java/app/organicmaps/routing/NavigationService.java

```diff
@@ -210,16 +210,21 @@ public void onLocationUpdated(@NonNull Location location)
     if (!routingController.isNavigating())
       return;
 
+    final String[] turnNotifications = Framework.nativeGenerateNotifications();
+    if (turnNotifications != null)
+      TtsPlayer.INSTANCE.playTurnNotifications(getApplicationContext(), turnNotifications);
+
     // TODO: consider to create callback mechanism to transfer 'ROUTE_IS_FINISHED' event from
     // the core to the platform code (https://github.com/organicmaps/organicmaps/issues/3589),
     // because calling the native method 'nativeIsRouteFinished'
     // too often can result in poor UI performance.
+    // This check should be done after playTurnNotifications() to play the last turn notification.
     if (Framework.nativeIsRouteFinished())
+    {
       routingController.cancel();
-
-    final String[] turnNotifications = Framework.nativeGenerateNotifications();
-    if (turnNotifications != null)
-      TtsPlayer.INSTANCE.playTurnNotifications(getApplicationContext(), turnNotifications);
+      stopSelf();
```

Comment:
Does it start next time properly after this call?

Reply:
Yes. NavigationService is always started by MwmActivity when navigation starts:

https://github.com/organicmaps/organicmaps/blob/bf63522cd4a0c20cf66fd7d232afda58350214ed/android/app/src/main/java/app/organicmaps/MwmActivity.java#L1511-L1527

The startService()/stopService()/stopSelf() operation are idempotent.